programs/stake/src/stake_state.rs

```diff
@@ -263,54 +263,81 @@ impl Delegation {
+            history.and_then(|history| {
+                history
+                    .get(&self.activation_epoch)
+                    .map(|cluster_stake_at_activation_epoch| {
+                        (
+                            history,
+                            self.activation_epoch,
+                            cluster_stake_at_activation_epoch,
+                        )
+                    })
+            })
+        {
+            // target_epoch > self.activation_epoch
 
-            // loop from my activation epoch until the current epoch
-            //   summing up my entitlement
+            // loop from my activation epoch until the target epoch summing up my entitlement
+            // current stake is updated using its previous epoch's cluster stake
+            let mut current_epoch;
+            let mut current_effective_stake = 0;
             loop {
-                if entry.activating == 0 {
+                current_epoch = prev_epoch + 1;
+                // if there is no activating stake at prev epoch, we should have been fully effective at this moment
+                if prev_cluster_stake.activating == 0 {
                     break;
                 }
+
                 // how much of the growth in stake this account is
                 //  entitled to take
-                let weight = (self.stake - effective_stake) as f64 / entry.activating as f64;
-
-                // portion of activating stake in this epoch I'm entitled to
-                effective_stake +=
-                    ((weight * entry.effective as f64 * self.warmup_cooldown_rate) as u64).max(1);
-
-                if effective_stake >= self.stake {
-                    effective_stake = self.stake;
+                let remaining_activating_stake = self.stake - current_effective_stake;
+                let weight =
+                    remaining_activating_stake as f64 / prev_cluster_stake.activating as f64;
+
+                // portion of newly effective cluster stake I'm entitled to at current epoch
+                let newly_effective_cluster_stake =
+                    prev_cluster_stake.effective as f64 * self.warmup_cooldown_rate;
+                let newly_effective_stake =
+                    ((weight * newly_effective_cluster_stake) as u64).max(1);
+
+                current_effective_stake += newly_effective_stake;
+                if current_effective_stake >= self.stake {
+                    current_effective_stake = self.stake;
                     break;
                 }
 
-                next_epoch += 1;
-                if next_epoch >= epoch || next_epoch >= self.deactivation_epoch {
+                if current_epoch >= target_epoch || current_epoch >= self.deactivation_epoch {
                     break;
                 }
-                if let Some(next_entry) = history.get(&next_epoch) {
-                    entry = next_entry;
+                if let Some(current_cluster_stake) = history.get(&current_epoch) {
```

Comment:
The whole various naming makes us to read this line like this even without the context:

> `history` is something which returns something historical (indexed by epoch) which is related to cluster-wide stake information for given epoch. Maybe some kind of totals of stakes. And we're getting new one for current_epoch and swap with prev for next iteration.